localstack/services/s3/cors.py

```diff
@@ -0,0 +1,250 @@
+from localstack.aws.protocol.serializer import gen_amzn_requestid_long
+from localstack.aws.protocol.service_router import get_service_catalog
+from localstack.constants import S3_VIRTUAL_HOSTNAME
+from localstack.http import Request, Response
+from localstack.services.s3.models import BucketCorsIndex
+from localstack.services.s3.utils import S3_VIRTUAL_HOSTNAME_REGEX
+
+_s3_virtual_host_regex = re.compile(S3_VIRTUAL_HOSTNAME_REGEX)
+FAKE_HOST_ID = "9Gjjt1m+cjU4OPvX9O9/8RuvnG41MRb/18Oux2o5H5MY7ISNTlXN+Dz9IG62/ILVxhAGI0qyPfg="
+
+# TODO: refactor those to expose the needed methods maybe in another way that both can import
+add_default_headers = CorsResponseEnricher.add_cors_headers
+is_origin_allowed_default = CorsEnforcer.is_cors_origin_allowed
+
+
+class S3CorsHandler(Handler):
+
+    bucket_cors_index: BucketCorsIndex
+
+    def __init__(self):
+        self.bucket_cors_index = BucketCorsIndex()
+        self._service = get_service_catalog().get("s3")
+        self._s3_op_router = RestServiceOperationRouter(self._service)
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        self.handle_cors(chain, context, response)
+
+    def pre_parse_s3_request(self, context: RequestContext) -> Tuple[bool, Optional[str]]:
+        is_s3: bool
+        bucket_name: str
+
+        path = context.request.path
+        host = context.request.host
+
+        # first, try to figure out best-effort whether the request is an s3 request
+        if host.startswith(S3_VIRTUAL_HOSTNAME):
+            is_s3 = True
+            bucket_name = path.split("/")[1]
+        # try to extract the bucket from the hostname (the "in" check is a minor optimization
+        elif ".s3" in host and (match := _s3_virtual_host_regex.match(host)):
+            is_s3 = True
+            bucket_name = match.group(3)
+        # otherwise we're not sure, and whether it's s3 depends on whether the bucket exists. check later
+        else:
+            is_s3 = False
+            bucket_name = path.split("/")[1]
+
+        existing_buckets = self.bucket_cors_index.buckets
+        if bucket_name not in existing_buckets:
+            return is_s3, None
+
+        return True, bucket_name
+
+    def handle_cors(self, chain: HandlerChain, context: RequestContext, response: Response):
+        """
+        Handle CORS for S3 requests. S3 CORS rules can be configured.
+        https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html
+        https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html
+        """
+
+        # this is used with the new ASF S3 provider
+        # although, we could use it to pre-parse the request and set the context to move the service name parser
+        if config.LEGACY_S3_PROVIDER or config.DISABLE_CUSTOM_CORS_S3:
+            return
+
+        is_s3, bucket_name = self.pre_parse_s3_request(context)
+
+        if not is_s3 and not bucket_name:
```

Comment:
Shouldn't this be an `or` condition?
The current condition would only let the default handler continue if it is _not_ S3 _and_ there is no bucket name (which would in my understanding be equal to just `if not is_s3`).
However, it seems like the default handler should continue for both cases, when it's not S3 but also if it is S3 but there is no bucket name?

Reply:
Right, this is a leftover from the multiple iterations we had with Thomas on the `pre_parse_s3_request` function.
`if not is_s3` will do the same, as the function cannot return `is_s3 = False` with a `bucket_name`. 
The default handler should not continue if it's S3 and there's no bucket name, there are specific errors in case of an `OPTIONS` request, it's a specific path in the flowchart.